File: electron/services/dependency-installer/interfaces/IProgressReporter.ts
// SPDX-License-Identifier: MIT OR Apache-2.0
// Copyright (c) 2025 Guilherme Ferrari Brescia

/**
 * Progress status types
 */
export type ProgressStatus =
  | "checking"
  | "downloading"
  | "installing"
  | "completed"
  | "error";

/**
 * Progress update information
 */
export interface InstallProgress {
  dependency: string;
  status: ProgressStatus;
  progress?: number;
  message: string;
  error?: string;
}

/**
 * Interface for progress reporting
 */
export interface IProgressReporter {
  report(progress: InstallProgress): void;
  onProgress(callback: (progress: InstallProgress) => void): void;
  removeProgressListener(callback: (progress: InstallProgress) => void): void;

  // Helper methods for specific status reports
  reportChecking(dependency: string, message: string): void;
  reportDownloading(
    dependency: string,
    message: string,
    progress?: number
  ): void;
  reportInstalling(dependency: string, message: string): void;
  reportCompleted(dependency: string, message: string): void;
  reportError(dependency: string, message: string, error?: string): void;
}

File: scripts/python/lora_training/interfaces/i_progress_reporter.py
# SPDX-License-Identifier: MIT OR Apache-2.0
"""
Progress reporter interface
"""

from abc import ABC, abstractmethod
from typing import Optional


class IProgressReporter(ABC):
    """Interface for progress reporting."""
    
    @abstractmethod
    def report_progress(self, current_step: int, total_steps: int, message: str, 
                       phase: Optional[str] = None) -> None:
        """Report training progress."""
        pass
    
    @abstractmethod
    def report_completion(self, message: str = "Training completed successfully") -> None:
        """Report training completion."""
        pass
    
    @abstractmethod
    def report_error(self, error_message: str) -> None:
        """Report training error."""
        pass 